### PR TITLE
pkg/registry/rbac/*,plugin/pkg/auth/authorizer/*: wire context

### PR DESCRIPTION
**Pull request by s-urbaniak**
#### What type of PR is this?

/kind cleanup

#### What this PR does / why we need it:

Currently, RBAC registry adapters create a request context on the fly but don't reuse the existing request context.

Specifically, context is wired in:
- ConfirmEscalation
- RBAC registries and their included adapters

This fixes it by wiring the context.

#### Which issue(s) this PR fixes:

N/A

#### Does this PR introduce a user-facing change?

```release-note
NONE
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs

```

### CODE REVIEW

File: pkg/auth/authorizer/abac/abac_test.go [hunk 1]
```diff
@@ -26,6 +26,8 @@ import (
 	"k8s.io/apimachinery/pkg/runtime"
 	"k8s.io/apiserver/pkg/authentication/user"
 	"k8s.io/apiserver/pkg/authorization/authorizer"
+	genericapirequest "k8s.io/apiserver/pkg/endpoints/request"
```
Comment (enj) [inline pkg/auth/authorizer/abac/abac_test.go:29]: Undo all new imports of this (applies to multiple files) and use `context.TODO()` in places where you do not have a context.

File: pkg/registry/rbac/validation/rule.go [hunk 2]
```diff
@@ -22,31 +22,31 @@ import (
...
-
 	rbacv1 "k8s.io/api/rbac/v1"
 	utilerrors "k8s.io/apimachinery/pkg/util/errors"
 	"k8s.io/apimachinery/pkg/util/sets"
 	"k8s.io/apiserver/pkg/authentication/serviceaccount"
 	"k8s.io/apiserver/pkg/authentication/user"
 	genericapirequest "k8s.io/apiserver/pkg/endpoints/request"
 	"k8s.io/component-helpers/auth/rbac/validation"
+	"k8s.io/klog/v2"
+
```
Comment (enj) [inline pkg/registry/rbac/validation/rule.go:33]:
```suggestion
```

File: pkg/registry/rbac/validation/rule.go [hunk 3]
```diff
@@ -100,24 +100,24 @@ func NewDefaultRuleResolver(roleGetter RoleGetter, roleBindingLister RoleBinding
...
 type ClusterRoleBindingLister interface {
-	ListClusterRoleBindings() ([]*rbacv1.ClusterRoleBinding, error)
+	ListClusterRoleBindings(ctx context.Context) ([]*rbacv1.ClusterRoleBinding, error)
 }
 
-func (r *DefaultRuleResolver) RulesFor(user user.Info, namespace string) ([]rbacv1.PolicyRule, error) {
+func (r *DefaultRuleResolver) RulesFor(ctx context.Context, user user.Info, namespace string) ([]rbacv1.PolicyRule, error) {
 	visitor := &ruleAccumulator{}
-	r.VisitRulesFor(user, namespace, visitor.visit)
+	r.VisitRulesFor(ctx, user, namespace, visitor.visit)
```
Comment (enj) [inline pkg/registry/rbac/validation/rule.go:120]: `visitor.visit` should take context as well.

**Review by enj — CHANGES_REQUESTED** [review]

File: plugin/pkg/auth/authorizer/rbac/rbac.go [hunk 4]
```diff
@@ -75,7 +75,7 @@ func (v *authorizingVisitor) visit(source fmt.Stringer, rule *rbacv1.PolicyRule,
 func (r *RBACAuthorizer) Authorize(ctx context.Context, requestAttributes authorizer.Attributes) (authorizer.Decision, string, error) {
 	ruleCheckingVisitor := &authorizingVisitor{requestAttributes: requestAttributes}
 
-	r.authorizationRuleResolver.VisitRulesFor(requestAttributes.GetUser(), requestAttributes.GetNamespace(), ruleCheckingVisitor.visit)
+	r.authorizationRuleResolver.VisitRulesFor(ctx, requestAttributes.GetUser(), requestAttributes.GetNamespace(), ruleCheckingVisitor.visit)
```
[the remark this answers is not in the document]
Comment (s-urbaniak) [reply plugin/pkg/auth/authorizer/rbac/rbac.go:78]: I didn't find traces where it is used below but I can add 👍

[the remark this answers is not in the document]
Comment (s-urbaniak) [reply plugin/pkg/auth/authorizer/rbac/rbac.go:78]: done

File: pkg/auth/authorizer/abac/abac_test.go (hunk 1 above)
Comment (s-urbaniak) [reply pkg/auth/authorizer/abac/abac_test.go:29]: done

File: pkg/registry/rbac/validation/rule.go (hunk 2 above)
Comment (s-urbaniak) [reply pkg/registry/rbac/validation/rule.go:33]: done

File: pkg/registry/rbac/validation/rule.go (hunk 3 above)
Comment (s-urbaniak) [reply pkg/registry/rbac/validation/rule.go:120]: done

**Comment by s-urbaniak** [comment]
@enj thank you for the swift review 👍 PTAL, addressed comments.

File: plugin/pkg/auth/authorizer/rbac/rbac.go [hunk 5]
```diff
@@ -60,7 +60,7 @@ type authorizingVisitor struct {
 	errors  []error
 }
 
-func (v *authorizingVisitor) visit(source fmt.Stringer, rule *rbacv1.PolicyRule, err error) bool {
+func (v *authorizingVisitor) visit(ctx context.Context, source fmt.Stringer, rule *rbacv1.PolicyRule, err error) bool {
```
Comment (enj) [inline plugin/pkg/auth/authorizer/rbac/rbac.go:63]: > I didn't find traces where it is used below but I can add 👍

These need to check the context for cancelation and terminate the visitor early in that case.

File: plugin/pkg/auth/authorizer/rbac/subject_locator.go [hunk 6]
```diff
@@ -18,21 +18,24 @@ limitations under the License.
 package rbac
 
 import (
+	"context"
+
 	rbacv1 "k8s.io/api/rbac/v1"
 	utilerrors "k8s.io/apimachinery/pkg/util/errors"
 	"k8s.io/apiserver/pkg/authentication/user"
 	"k8s.io/apiserver/pkg/authorization/authorizer"
+
```
[the remark this answers is not in the document]
Comment (s-urbaniak) [reply plugin/pkg/auth/authorizer/rbac/subject_locator.go]: ah, sorry, fixed 👍

File: pkg/registry/rbac/role/registry.go [hunk 7]
```diff
@@ -62,6 +63,6 @@ type AuthorizerAdapter struct {
 }
 
 // GetRole returns the corresponding Role by name in specified namespace
-func (a AuthorizerAdapter) GetRole(namespace, name string) (*rbacv1.Role, error) {
-	return a.Registry.GetRole(genericapirequest.WithNamespace(genericapirequest.NewContext(), namespace), name, &metav1.GetOptions{})
+func (a AuthorizerAdapter) GetRole(ctx context.Context, namespace, name string) (*rbacv1.Role, error) {
+	return a.Registry.GetRole(genericapirequest.WithNamespace(ctx, namespace), name, &metav1.GetOptions{})
```
Comment (deads2k) [inline pkg/registry/rbac/role/registry.go:67]: What should happen if the context comes in with a different namespace defined?

File: pkg/registry/rbac/rolebinding/registry.go [hunk 8]
```diff
@@ -61,8 +61,8 @@ type AuthorizerAdapter struct {
 	Registry Registry
 }
 
-func (a AuthorizerAdapter) ListRoleBindings(namespace string) ([]*rbacv1.RoleBinding, error) {
-	list, err := a.Registry.ListRoleBindings(genericapirequest.WithNamespace(genericapirequest.NewContext(), namespace), &metainternalversion.ListOptions{})
+func (a AuthorizerAdapter) ListRoleBindings(ctx context.Context, namespace string) ([]*rbacv1.RoleBinding, error) {
+	list, err := a.Registry.ListRoleBindings(genericapirequest.WithNamespace(ctx, namespace), &metainternalversion.ListOptions{})
```
Comment (deads2k) [inline pkg/registry/rbac/rolebinding/registry.go:65]: same question about conflicting namespaces here.

File: pkg/registry/rbac/validation/rule.go [hunk 9]
```diff
@@ -188,26 +187,26 @@ func (r *DefaultRuleResolver) VisitRulesFor(user user.Info, namespace string, vi
 			if !applies {
 				continue
 			}
-			rules, err := r.GetRoleReferenceRules(clusterRoleBinding.RoleRef, "")
+			rules, err := r.GetRoleReferenceRules(ctx, clusterRoleBinding.RoleRef, "")
 			if err != nil {
-				if !visitor(nil, nil, err) {
+				if !visitor(ctx, nil, nil, err) {
```
Comment (deads2k) [inline pkg/registry/rbac/validation/rule.go:192]: didn't @enj ask for cancellation to be honored?

File: pkg/registry/rbac/role/registry.go (hunk 7 above)
Comment (enj) [reply pkg/registry/rbac/role/registry.go:67]: Hmm what does this look like when a webhook authorizer is invoked via SAR?

**Comment by enj** [comment]
> > Currently, RBAC registry adapters create a request context on the fly but don't reuse the existing request context.
> 
> > Specifically, context is wired in:
> 
> > ConfirmEscalation
> > RBAC registries and their included adapters
> > This fixes it by wiring the context.
> 
> I don't see anything actually using this context in this PR. Why change the interface?

Calls to the rule resolver such as [this](https://github.com/kubernetes/kubernetes/pull/113971/files#diff-8cf30f4b2a919f62a6adf1b8b86b817c4486644d43703af7b8e570d0b05d3e47R92) hit the underlying registry so these will support cancellation now.

**Comment by deads2k** [comment]
> Calls to the rule resolver such as [this](https://github.com/kubernetes/kubernetes/pull/113971/files#diff-8cf30f4b2a919f62a6adf1b8b86b817c4486644d43703af7b8e570d0b05d3e47R92) hit the underlying registry so these will support cancellation now.

I don't see the code that actually checks and honors the cancel or passes it to a live client.  Listers don't honor cancel.

**Comment by deads2k** [comment]
It *could* honor cancellation, but I don't this PR wires it up.  I'm interested in the motivation for adding this.  Are we fixing a problem and thus need a test?  Does someone actually produce another implementation of this somewhere and they need cancellation or perhaps worse, want to use the context as a means to wire a parameter from one part of the stack (maybe a filter or http handler?) down to the rules resolver?

**Comment by deads2k** [comment]
> Are we fixing a problem and thus need a test? Does someone actually produce another implementation of this somewhere and they need cancellation or perhaps worse, want to use the context as a means to wire a parameter from one part of the stack (maybe a filter or http handler?) down to the rules resolver?

I found the linked PR.  Turns out this is desired for side-band argument passing: https://github.com/kcp-dev/kcp/blob/ee2d2d15000596539b15c402f678b164bae0881b/pkg/authorization/toplevel_org_authorizer.go#L90

I wouldn't recommend doing such a thing, but contexts are generally good for kube.  Though seeing abuse like this, it may be good for kube to start controlling what keys are legal to add to a context so that other areas of kube code don't start growing a similar technique for argument passing.

I think the cancellation handling is outstanding the conflicting namespace comments need resolution of some kind since you're constructing it.

File: pkg/registry/rbac/validation/rule.go (hunk 9 above)
Comment (s-urbaniak) [reply pkg/registry/rbac/validation/rule.go:192]: yes, indeed, this is WIP on my local branch, let me switch the title to that status as well 👍

**Comment by dims** [comment]
This PR has the label work-in-progress, please revisit to see if you still need this, please close if not

**Comment by s-urbaniak** [comment]
@dims yes, i am still working on this

File: plugin/pkg/auth/authorizer/rbac/rbac.go (hunk 5 above)
Comment (s-urbaniak) [reply plugin/pkg/auth/authorizer/rbac/rbac.go:63]: @enj addressed in latest commit https://github.com/kubernetes/kubernetes/pull/113971/commits/800bed1cec1e1c0b58d5382fd85f883d7bb4a5b0, could you PTAL?

**Comment by s-urbaniak** [comment]
@deads2k @enj addressed comments, ready for a next iteration, could you please PTAL 🙏

File: pkg/registry/rbac/rolebinding/registry.go (hunk 8 above)
Comment (s-urbaniak) [reply pkg/registry/rbac/rolebinding/registry.go:65]: https://github.com/kubernetes/kubernetes/pull/113971#discussion_r1053311491 applies here as well

**Pull request closed without merging**